File: RealTimeGisWithSignalR/realTimePoints.js
dojo.require("esri.map");
dojo.require("esri.layers.agstiled");
dojo.require("esri.toolbars.draw");

var map, tb;

function init() {

    // Initiate map with the "Streets" basemap.
    map = new esri.Map("map", {
        basemap: "streets",
        center: [-25.312, 34.307],
        zoom: 3
    });

    dojo.connect(map, "onLoad", initToolbar);
}

var realTimePoints;
var layersColors = [];

$(function () {

    // Create Proxy to SignalR hub
    realTimePoints = $.connection.realTimePoints;

    // This function is callable from the server
    realTimePoints.client.addPoint = function addPoint(cid, x, y) {

        // Create point object and set its geometry
        var p = new esri.geometry.Point(x, y, new esri.SpatialReference({ wkid: 102100 }));

        // Get the layer of remote user to which the line will be added
        var gLayer = map.getLayer(cid);

        // Set up symbol and color of point
        // Get color from the layersColors list
        var symbol = new esri.symbol.SimpleMarkerSymbol();
        var clr = layersColors[layersColors.indexOf(cid) + 1];
        symbol.setColor(new dojo.Color(clr));

        // Add point to graphics layer
        gLayer.add(new esri.Graphic(p, symbol));
    };

    // This function is callable from the server
    realTimePoints.client.addPolyline = function addPolyline(cid, paths) {

        // Create line object and set its geometry
        var line = new esri.geometry.Polyline(new esri.SpatialReference({ wkid: 102100 }));
        line.paths = paths;

        // Get the layer of remote user to which the line will be added
        var gLayer = map.getLayer(cid);

        // Set up symbol and color of line
        // Get color from the layersColors list
        var symbol = new esri.symbol.SimpleLineSymbol();
        var clr = layersColors[layersColors.indexOf(cid) + 1];
        symbol.setColor(new dojo.Color(clr));

        // Add line to graphics layer
        gLayer.add(new esri.Graphic(line, symbol));
    };

    // This function is callable from the server
    realTimePoints.client.addLayer = function addLayer(cid, color) {

        // Create new layer
        var gLayer = new esri.layers.GraphicsLayer();
        gLayer.id = cid;

        // Keep track of layer names and their colors.
        layersColors.push(cid);
        layersColors.push(color);

        // Add layer
        map.addLayer(gLayer);

        // Update number of layers counter
        $("#graphicsLayersCount").empty();
        $("#graphicsLayersCount").append("<p>Number of graphics layers: " + map.graphicsLayerIds.length + "</p>");
        
        // Set the default symbols for drawing
        if ($.connection.hub.id === cid) {

            var clr = layersColors[layersColors.indexOf(cid) + 1];

            // Point symbol
            var symbol = new esri.symbol.SimpleMarkerSymbol();
            symbol.setColor(new dojo.Color(clr));
            tb.markerSymbol = symbol;

            // Line symbol
            symbol = new esri.symbol.SimpleLineSymbol();
            symbol.setColor(new dojo.Color(clr));
            tb.lineSymbol = symbol;
        }
    };

    // This function is callable from the server
    realTimePoints.client.removeLayer = function removeLayer(cid) {
                
        if ($.connection.hub.id !== cid) {

            // Get layer with specific cid (Connection ID)
            var gLayer = map.getLayer(cid);

            // Remove layer
            map.removeLayer(gLayer);

            // Update number of layers counter
            $("#graphicsLayersCount").empty();
            $("#graphicsLayersCount").append("<p>Number of graphics layers: " + map.graphicsLayerIds.length + "</p>");
        }
    };
    
    // This function is callable from the server
    realTimePoints.client.updataGraphicsLayersLegend =
        function updataGraphicsLayersLegend(layersNames, layersColors) {

            // This function updates the legend

            $("#legendDiv").empty();
            $("#legendDiv").append("<p><b>Legend</b></p>");

            for (var i = 0; i < layersNames.length; i++) {

                $("#legendDiv").append("<p><font color='" + layersColors[i] + "'>" + layersNames[i] + "</font></p>");
            }
        };

    // This function is callable from the server
    realTimePoints.client.updateOwnUserName = function updateOwnUserName(initialUserName) {
        // This function sets the user's initial username which is assigned by the server.
        $("#txtUserName").val(initialUserName);
    };

    // Everything is ready, now start the connection
    $.connection.hub.start();

});

function btnUpdateUserName_Click() {
    realTimePoints.server.updateUserName($("#txtUserName").val());
}

function initToolbar(map) {
    // Initiate Drawing toolbar
    tb = new esri.toolbars.Draw(map);
    dojo.connect(tb, "onDrawEnd", addGraphic);
}

function addGraphic(geometry) {
    var type = geometry.type;
    if (type === "point" || type === "multipoint") {

        var symbol = new esri.symbol.SimpleMarkerSymbol();
        var clr = layersColors[layersColors.indexOf($.connection.hub.id) + 1];
        symbol.setColor(new dojo.Color(clr));
        // Notify the server about the point
        realTimePoints.server.addPoint(geometry.x, geometry.y);
    }
    else if (type === "line" || type === "polyline") {

        var symbol = new esri.symbol.SimpleLineSymbol();
        var clr = layersColors[layersColors.indexOf($.connection.hub.id) + 1];
        symbol.setColor(new dojo.Color(clr));
        // Notify the server about the line
        realTimePoints.server.addPolyline(geometry.paths);
    }
    else {
        symbol = tb.fillSymbol;
    }
}

dojo.ready(init);

File: RealTimeGisWithSignalR/RealTimePoints.cs
using System.Collections.Generic;
using System.Linq;
using Microsoft.AspNet.SignalR;

namespace RealTimeGisWithSignalR.Test
{
    public class RealTimePoints : Hub
    {
        private static int _ConnectionCount;

        // Dictionary of usernames identified by connection ID
        public static Dictionary<string, string> _UserNames = new Dictionary<string, string>();
        // Dictionary of layers  identified by connection ID
        public static Dictionary<string, string> _LayersLegend = new Dictionary<string, string>();

        public static List<string> _Colors = new List<string>() { "red", "blue", "green", "black", "grey", "purple", "fuchsia", "silver", "lime", "olive", "yellow", "navy", "teal", "aqua", "maroon" };

        // Public method callable from clients
        public void AddPoint(double x, double y)
        {
            // Notify clients to add point
            Clients.All.addPoint(Context.ConnectionId, x, y);
        }

        // Public method callable from clients
        public void AddPolyline(double[][][] paths)
        {
            // Notify clients to add line
            Clients.All.addPolyline(Context.ConnectionId, paths);
        }

        // Public method callable from clients
        public void AddPolygon(double[][][] rings)
        {
            // Notify clients to add polygon
            Clients.All.addPolygon(Context.ConnectionId, rings);
        }

        // Public method callable from clients
        public void UpdateUserName(string userName)
        {
            // Update username
            if (_UserNames.Keys.Contains(Context.ConnectionId))
            {
                _UserNames[Context.ConnectionId] = userName;
            }

            // Pass updates to all clients
            Clients.All.updataGraphicsLayersLegend(
                _UserNames.Values.ToArray(),
                _LayersLegend.Values.ToArray());
        }

        // Handle new clients connected
        public override System.Threading.Tasks.Task OnConnected()
        {
            // Add existing graphics layers to caller (new client)
            int counter = 0;
            foreach (string graphicsLayer in _UserNames.Keys)
            {
                string clr = _Colors[counter];
                Clients.Caller.addLayer(graphicsLayer, clr);
                counter += 1;
            }

            // Notify all clients (including new client) to add new graphics layer for new client.
            string color = _Colors[_ConnectionCount];
            Clients.All.addLayer(Context.ConnectionId, color);

            // Generate initial username
            string initialUserName = "user" + Context.ConnectionId.Substring(Context.ConnectionId.Length - 3);

            // Pass initial user name to caller
            Clients.Caller.updateOwnUserName(initialUserName);

            // Add username and color to legend
            _UserNames.Add(Context.ConnectionId, initialUserName);
            _LayersLegend.Add(Context.ConnectionId, color);

            // Notify all clients to update their legend.
            Clients.All.updataGraphicsLayersLegend(_UserNames.Values.ToArray(), _LayersLegend.Values.ToArray());

            // Update connections counter
            _ConnectionCount += 1;

            return base.OnConnected();
        }

        // Handle clients disconnected
        public override System.Threading.Tasks.Task OnDisconnected()
        {
            // Update connections counter
            _ConnectionCount -= 1;

            // Notify all clients to remove layer.
            Clients.All.removeLayer(Context.ConnectionId);

            // Remove client's info
            _UserNames.Remove(Context.ConnectionId);
            _LayersLegend.Remove(Context.ConnectionId);

            // Notify all clients to update their legend.
            Clients.All.updataGraphicsLayersLegend(
                _UserNames.Values.ToArray(),
                _LayersLegend.Values.ToArray());

            return base.OnDisconnected();
        }
    }
}